### ISSUE TITLE
Support SSO to GitHub OIDC with specific organization

### ISSUE THREAD

**Issue by 08d2** [matrix-org/synapse#12238]
This is a new issue based on [this comment](https://github.com/matrix-org/synapse/issues/9356#issuecomment-1069408999). I'm trying to configure SSO/OIDC to GitHub such that only users in a specific organization are permitted. But this doesn't seem to be possible.

**Comment by DMRobertson** [comment]
If someone wants to take a crack at this, it looks like key places to change this are:

1) the structure defining an attribute requirement

https://github.com/matrix-org/synapse/blob/f44d729d4ccae61bc0cdd5774acb3233eb5f7c13/synapse/config/sso.py#L31-L44

2) the logic which checks that requirement

https://github.com/matrix-org/synapse/blob/3f7cfbc9e5e82d17cd7782adcbe78e239aae5691/synapse/handlers/sso.py#L1029-L1059

**Comment by clokep** [comment]
If you add the [`read:org` scope](https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps) to the [list of scopes that Synapse requests](https://github.com/matrix-org/synapse/blob/b1989ced00cc0bc6214bfd1a393c7e8f8eda660c/docs/sample_config.yaml#L1896-L1897) does that give the additional information you would need to filter on?

**Comment by 08d2** [comment]
@clokep 

> If you add the [`read:org` scope](https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps) to the [list of scopes that Synapse requests](https://github.com/matrix-org/synapse/blob/b1989ced00cc0bc6214bfd1a393c7e8f8eda660c/docs/sample_config.yaml#L1896-L1897) does that give the additional information you would need to filter on?

Nope, as the returned user info doesn't include organizations directly, it provides a URL — which can be successfully retrieved with `read:org`. 

---

@DMRobertson 

> I don't think we're keen to add any logic or integration specific to GitHub.

Consider the e.g. [Grafana integration with GitHub](https://grafana.com/docs/grafana/latest/auth/github/#enable-github-in-grafana), which accepts `team_ids` and `allowed_organizations` parameters and satisfies them with GitHub-specific code. As a user, auth constraints like this are requirements for an SSO integration to be useful.
